src/ring_theory/bundled_subring.lean

```diff
@@ -0,0 +1,736 @@
+/-
+Copyright (c) 2020 Ashvni Narayanan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors : Ashvni Narayanan
+-/
+
+import group_theory.subgroup
+import ring_theory.subsemiring
+
+/-!
+# Subrings
+
+Let `R` be a ring. This file defines the "bundled" subring type `subring R`, a type
+whose terms correspond to subrings of `R`. This is the preferred way to talk
+about subrings in mathlib. Unbundled subrings (`s : set R` and `is_subring s`)
+are not in this file, and they will ultimately be deprecated.
+
+We prove that subrings are a complete lattice, that you can `map` (pushforward) and
+`comap` (pull back) them along ring homomorphisms.
+
+We define the `closure` construction from `set R` to `subgroup R`, sending a subset of `R`
+to the subring it generates, and prove that it is a Galois insertion.
+
+## Main definitions
+
+Notation used here:
+
+`(R : Type u) [ring R] (S : Type u) [ring S] (f g : R →+* S)`
+`(A : subring R) (B : subring S) (s : set R)`
+
+* `subring R` : the type of subrings of a ring `R`.
+
+* `instance : complete_lattice (subring R)` : the complete lattice structure on the subrings.
+
+* `subring.closure` :  subring closure of a set, i.e., the smallest subring that includes the set.
+
+* `subring.gi` : `closure : set M → subring M` and coercion `coe : subring M → set M`
+  form a `galois_insertion`.
+
+* `closure s` : the minimal subring that includes the set `s`
+
+* `comap f B : subring A` : the preimage of a subring `B` along the ring homomorphism `f`
+
+* `map f A : subring B` : the image of a subring `A` along the ring homomorphism `f`.
+
+* `prod A B : subring (R × S)` : the product of subring
+
+* `f.range : subring B` : the range of the ring homomorphism `f`.
+
+* `eq_locus f g : subring R` : given ring homomorphisms `f`, `g`,
+     the subring of `R` where `f x = g x`
+
+## Implementation notes
+
+Lattice inclusion (e.g. `≤` and `⊓`) is used rather than set notation (`⊆` and `∩`), although
+`∈` is defined as membership of a subring's underlying set.
+
+## Tags
+subring, subrings
+-/
+
+open_locale big_operators
+universes u v w
+
+open group
+variables {R : Type u} {S : Type v} {T : Type w} [ring R] [ring S] [ring T]
+
+set_option old_structure_cmd true
+
+/-- `subring R` is the type of subrings of `R`. A subring of `R` is a subset `s` that is a multiplicative
+submonoid and an additive subgroup. Note in particular that it shares the same 0 and 1 as R -/
+structure subring (R : Type u) [ring R] extends submonoid R, add_subgroup R
```

Comment:
Perhaps we should also have a construction taking a `subsemiring` of a `ring`, along with evidence that `-1` is in the subsemiring, and produce a `subring`?

(I don't think this is essential for this PR.)